drivers/platform/xilinx/xilinx_spi_pl.c

```diff
@@ -0,0 +1,417 @@
+	ldesc->mode = param->mode;
+	xdesc->base_addr = cfg->BaseAddress;
+	xdesc->disable_slaves_mask = (1 << cfg->NumSlaveBits) - 1;
+	xdesc->cs_asserted = 0;
+
+	err = _xil_spi_init_dev(xdesc);
+	if (IS_ERR_VALUE(err))
+		goto error;
+
+	ldesc->extra = xdesc;
+	*desc = ldesc;
+
+	return SUCCESS;
+
+error:
+	free(xdesc);
+	free(ldesc);
+
+	return err;
+}
+
+/* Remove allocated resources */
+static int32_t xil_spi_remove_pl(struct spi_desc *desc)
+{
+	if (!desc)
+		return -EINVAL;
+
+	free(desc->extra);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/* Fill fifo until fifo is full or no more data available */
+static inline uint32_t _wr_fifo(struct xspi_desc *xdesc, uint8_t *data,
+				uint32_t len)
+{
+	uint32_t i;
+	uint32_t to_write;
+	uint32_t tx_available;
+
+	tx_available = xdesc->fifo_depth - _read(xdesc, XSP_TFO_OFFSET);;
+	to_write = min(tx_available, len);
+	for (i = 0; i < to_write; ++i)
+		_write(xdesc, XSP_DTR_OFFSET, data[i]);
+
+	return to_write;
+}
+
+/* Read fifo until fifo is empty or no more buffer available to read into */
+static inline uint32_t _rd_fifo(struct xspi_desc *xdesc, uint8_t *data,
+				uint32_t len)
+{
+	uint32_t rx_available;
+	uint32_t to_read;
+	uint32_t tmp;
+	uint32_t i;
+
+	rx_available = _read(xdesc, XSP_RFO_OFFSET) + 1;
+	to_read = min(rx_available, len);
+
+	for (i = 0; i < to_read; ++i) {
+		tmp = _read(xdesc, XSP_DRR_OFFSET);
+		if (data)
+			data[i] = tmp;
+	}
+
+	return to_read;
+}
+
+/* If start is 0 stop transfer, else start it */
+static void inline _xil_spi_start_transfer(struct spi_desc *desc, uint8_t start)
+{
+	struct xspi_desc	*xdesc;
+
+	xdesc = desc->extra;
+	if (start) {
+		_write(xdesc, XSP_SSR_OFFSET, ~(1 << desc->chip_select));
+		_update(xdesc, XSP_CR_OFFSET, 0, XSP_CR_TRANS_INHIBIT_MASK);
+	} else {
+		_update(xdesc, XSP_CR_OFFSET, XSP_CR_TRANS_INHIBIT_MASK,
+			XSP_CR_TRANS_INHIBIT_MASK);
+		_write(xdesc, XSP_SSR_OFFSET, xdesc->disable_slaves_mask);
+	}
+	xdesc->cs_asserted = start;
+}
+
+
+/* SPI polled transfer */
+static int32_t xil_spi_write_and_read_pl(struct spi_desc *desc, uint8_t *data,
+		uint16_t bytes_number)
+{
+	struct xspi_desc	*xdesc;
+	uint32_t		tx;
+	uint32_t		rx;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	xdesc = desc->extra;
+	_update_mode(desc);
+
+	rx = 0;
+	tx = 0;
+	_xil_spi_start_transfer(desc, 1);
+	while (rx < bytes_number) {
+		if (tx < bytes_number)
+			tx += _wr_fifo(xdesc, data + tx, bytes_number - tx);
+		rx += _rd_fifo(xdesc, data + rx, bytes_number - rx);
+	}
+	_xil_spi_start_transfer(desc, 0);
+
+	return SUCCESS;
+}
+
+/* Read/write FIFOs to/from buffers and update it structure accordingly.
+ * Return */
+static inline uint8_t _transfer_and_update(struct xspi_desc *xdesc,
+		struct msg_iterator *it)
+{
+	uint32_t		i;
+	uint32_t		total;
+	uint32_t		len;
+	uint8_t			cs_change;
+
+	total = 0;
+	cs_change = 0;
+	do {
+		len = it->msg->bytes_number - it->i;
+		if (it->is_write)
+			i = _wr_fifo(xdesc, it->buff + it->i, len);
+		else if (it->buff)
```

Comment:
why this check? isn't it the se with _rd_fifo(xdesc, NULL, len); if it->buff == NULL ?

Reply:
nevermind this just saw that it->buff gets incremented with it->i